jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java

```diff
@@ -941,8 +941,118 @@ else if (percentile == 1.0)
+    }
+
+    /**
+     * CS304 Issue link: https://github.com/jOOQ/jOOL/issues/221
+     * Get a {@link Collector} that calculates the derived <code>PERCENTILE_DISC(percentile)</code> function given a specific ordering, producing multiple results.
+     */
+    public static <T, U extends Comparable<? super U>> Collector<T, ?, Seq<T>> percentileAllBy(double percentile, Function<? super T, ? extends U> function) {
+        return percentileAllBy(percentile, function, naturalOrder());
+    }
+
+    /**
+     * CS304 Issue link: https://github.com/jOOQ/jOOL/issues/221
+     * Get a {@link Collector} that calculates the derived <code>PERCENTILE_DISC(percentile)</code> function given a specific ordering, producing multiple results.
+     */
+    public static <T, U> Collector<T, ?, Seq<T>> percentileAllBy(double percentile, Function<? super T, ? extends U> function, Comparator<? super U> comparator) {
+        if (percentile < 0.0 || percentile > 1.0)
+            throw new IllegalArgumentException("Percentile must be between 0.0 and 1.0");
+
+        if (percentile == 0.0)
+            // If percentile is 0, this is the same as taking the items with the minimum value.
+            return minAllBy(function, comparator);
+        else if (percentile == 1.0)
+            // If percentile is 1, this is the same as taking the items with the maximum value.
+            return maxAllBy(function, comparator);
+
+        return Collector.of(
+                () -> new ArrayList<Tuple2<? extends T, U>>(),
+                (l, v) -> l.add(tuple(v, function.apply(v))),
+                (l1, l2) -> {
+                    l1.addAll(l2);
+                    return l1;
+                },
+                l -> {
+                    int size = l.size();
+
+                    if (size == 0)
+                        return Seq.empty();
+                    else if (size == 1)
+                        return Seq.of(l.get(0).v1);
+
+                    l.sort(Comparator.comparing(t -> t.v2, comparator));
+
+                    int left = 0;
+                    int right = (int) -Math.round(-(size * percentile + 0.5)) - 1;
+                    int mid;
+
+                    List<T> result = new ArrayList<>();
+                    U value = l.get(right).v2;
+                    // Search the first value equal to the value at the position of PERCENTILE by binary search
+                    while (left < right) {
+                        mid = left + (right - left) / 2;
+                        if (comparator.compare(l.get(mid).v2, value) < 0)
+                            left = mid + 1;
+                        else
+                            right = mid;
+                    }
+
+                    for (; left < size && comparator.compare(l.get(left).v2, value) == 0; left++)
+                        result.add(l.get(left).v1);
+                    return Seq.seq(result);
+                }
+        );
+    }
```

Comment:
Have you checked, is there any refactoring potential to have common logic between `percentileBy()` and `percentileAllBy()`?

Reply:
I mean, at least the collection to the accumulator seems to be the same, the difference being the finisher implementation.

Reply:
I don't like the similar structure of `percentileBy` and `percentileAllBy` too, but I didn't find a proper way to refactor these codes.

Reply:
Sure, no worries. I'll refactor it. Thanks for the effort!

Reply:
This should do the trick:

```java
private static <T, U, R> Collector<T, List<Tuple2<T, U>>, R> percentileCollector(
    Function<? super T, ? extends U> function,
    Comparator<? super U> comparator,
    Supplier<R> onEmpty,
    Function<T, R> onSingle,
    Function<List<Tuple2<T, U>>, R> finisher
) {
    return Collector.of(
        () -> new ArrayList<>(),
        (l, v) -> l.add(tuple(v, function.apply(v))),
        (l1, l2) -> {
            l1.addAll(l2);
            return l1;
        },
        l -> {
            int size = l.size();

            if (size == 0)
                return onEmpty.get();
            else if (size == 1)
                return onSingle.apply(l.get(0).v1);

            l.sort(Comparator.comparing(t -> t.v2, comparator));
            return finisher.apply(l);
        }
    );
}
```

Leading to e.g.

```java
public static <T, U> Collector<T, ?, Optional<T>> percentileBy(double percentile, Function<? super T, ? extends U> function, Comparator<? super U> comparator) {
    if (percentile < 0.0 || percentile > 1.0)
        throw new IllegalArgumentException("Percentile must be between 0.0 and 1.0");

    if (percentile == 0.0)
        return minBy(function, comparator);
    else if (percentile == 1.0)
        return collectingAndThen(maxAllBy(function, comparator), s -> s.findLast());
    else
        return percentileCollector(
            function,
            comparator,
            Optional::empty,
            Optional::of,
            l -> Optional.of(l.get(percentileIndex(percentile, l.size())).v1)
        );
}
```

Where 

```java
private static int percentileIndex(double percentile, int size) {

    // x.5 should be rounded down
    return (int) -Math.round(-(size * percentile + 0.5)) - 1;
}
```